src/app/features/search-bar/search-bar.component.ts

```diff
@@ -0,0 +1,250 @@
+import { Router } from '@angular/router';
+import { TODAY_TAG } from '../tag/tag.const';
+import { Project } from '../project/project.model';
+import { Tag } from '../tag/tag.model';
+import { ProjectService } from '../project/project.service';
+import { TagService } from '../tag/tag.service';
+import { Task } from '../tasks/task.model';
+import { blendInOutAnimation } from 'src/app/ui/animations/blend-in-out.ani';
+import { AnimationEvent } from '@angular/animations';
+import { SearchItem, SearchQueryParams } from './search-bar.model';
+import { getWorklogStr } from '../../util/get-work-log-str';
+
+@Component({
+  selector: 'search-bar',
+  templateUrl: './search-bar.component.html',
+  styleUrls: ['./search-bar.component.scss'],
+  animations: [blendInOutAnimation],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchBarComponent implements AfterViewInit, OnDestroy {
+  @Input() isElevated: boolean = false;
+  @Output() blurred: EventEmitter<any> = new EventEmitter();
+
+  @ViewChild('inputEl') inputEl!: ElementRef;
+  @ViewChild('searchForm') searchForm!: ElementRef;
+
+  T: typeof T = T;
+  isLoading$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  taskSuggestionsCtrl: FormControl = new FormControl();
+  filteredIssueSuggestions$: Observable<SearchItem[]> = new Observable();
+  isArchivedTasks: boolean = false;
+  tooManyResults: boolean = false;
+
+  private _attachKeyDownHandlerTimeout?: number;
+  private _tasks: Task[] = [];
+  private _projects: Project[] = [];
+  private _tags: Tag[] = [];
+  private _searchableItems: SearchItem[] = [];
+
+  constructor(
+    private _taskService: TaskService,
+    private _projectService: ProjectService,
+    private _tagService: TagService,
+    private _router: Router,
+  ) {
+    this._taskService.allTasks$.pipe(take(1)).subscribe((t) => (this._tasks = t));
+    this._projectService.list$.pipe(take(1)).subscribe((p) => (this._projects = p));
+    this._tagService.tags$.pipe(take(1)).subscribe((t) => (this._tags = t));
+
+    this._mapTasksToSearchItems();
+  }
+
+  private _mapTasksToSearchItems(): void {
+    this._searchableItems = this._tasks.map((task) => {
+      const item: SearchItem = {
+        id: task.id,
+        title: task.title.toLowerCase(),
+        taskNotes: task.notes.toLowerCase(),
+        projectId: task.projectId,
+        parentId: task.parentId,
+        tagIds: task.tagIds,
+        timeSpentOnDay: task.timeSpentOnDay,
+        createdOn: task.created,
+        issueType: task.issueType,
+        ctx: this._getContextIcon(task),
+      };
+      return item;
+    });
+  }
+
+  private _getContextIcon(task: Task) {
+    const context = this._getCtxForTaskSuggestion(task);
+    return {
+      ...context,
+      icon: (context as Tag).icon || (task.projectId && 'list'),
+    };
+  }
+
+  ngAfterViewInit(): void {
+    this._setUpFilter();
+
+    this._attachKeyDownHandlerTimeout = window.setTimeout(() => {
```

Comment:
Love that you thought about canceling the timeout :)

Reply:
😅 actually, that has been carried over form add-task-bar. Written by you :)
But would love to hear why it is good practice.

Reply:
makes good sense. thanks.